src/main/scala/mesosphere/marathon/health/MarathonHealthCheckManager.scala

```diff
@@ -115,30 +115,40 @@ class MarathonHealthCheckManager @Inject() (
   override def reconcileWith(appId: PathId): Future[Unit] =
     appRepository.currentVersion(appId) flatMap {
       case None => Future(())
-      case Some(app) => appHealthChecks.writeLock { ahcs =>
-        val versionTasksMap: Map[String, MarathonTask] =
-          taskTracker.getTasks(app.id).iterator.map { task => task.getVersion -> task }.toMap
-
-        // remove health checks for which the app version is not current and no tasks remain
-        // since only current version tasks are launched.
-        for {
-          (version, activeHealthChecks) <- ahcs(appId)
-          if version != app.version && !versionTasksMap.contains(app.id.toString)
-          activeHealthCheck <- activeHealthChecks
-        } {
-          remove(appId, version, activeHealthCheck.healthCheck)
+      case Some(app) =>
+        log.info(s"reconcile [$appId] with latest version [${app.version}]")
+
+        val tasks: Iterable[MarathonTask] = taskTracker.getTasks(app.id)
+        val activeAppVersions: Set[String] = tasks.iterator.map(_.getVersion).toSet + app.version.toString
+
+        val healthCheckAppVersions: Set[String] = appHealthChecks.writeLock { ahcs =>
```

Comment:
This not only computes the `healthCheckAppVersions` but also removes health checks. Please extract the remove operation.